sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/BlobDecryptionPolicy.java

```diff
@@ -76,68 +94,146 @@ public class BlobDecryptionPolicy implements HttpPipelinePolicy {
+                    /*
+                     * We will need to know the total size of the data to know when to finalize the decryption. If it
+                     * was not set originally with the intent of downloading the whole blob, update it here.
+                     * If there was no range set on the request, we skipped instantiating a BlobRange as we did not have
+                     * encryption data at the time. Instantiate now with a BlobRange that indicates a full blob.
+                     */
+                    EncryptedBlobRange encryptedRange = new EncryptedBlobRange(new BlobRange(0), encryptionData);
+                    encryptedRange.setAdjustedDownloadCount(
+                        Long.parseLong(responseHeaders.getValue(CONTENT_LENGTH)));
 
-                Flux<ByteBuffer> plainTextData = this.decryptBlob(encryptedDataString,
-                    httpResponse.getBody(), encryptedRange, padding);
+                    /*
+                     * We expect padding only if we are at the end of a blob and it is not a multiple of the encryption
+                     * block size. Padding is only ever present in track 1.
+                     */
+                    boolean padding = encryptionData.getEncryptionAgent().getProtocol().equals(ENCRYPTION_PROTOCOL_V1)
+                        && (encryptedRange.toBlobRange().getOffset()
+                        + encryptedRange.toBlobRange().getCount()
+                        > (blobSize(responseHeaders) - ENCRYPTION_BLOCK_SIZE));
 
-                return Mono.just(new BlobDecryptionPolicy.DecryptedResponse(httpResponse, plainTextData));
-            } else {
-                return Mono.just(httpResponse);
-            }
-        });
+                    Flux<ByteBuffer> plainTextData = this.decryptBlob(httpResponse.getBody(), encryptedRange, padding,
+                        encryptionData, httpResponse.getRequest().getUrl().toString());
+
+                    return Mono.just(new BlobDecryptionPolicy.DecryptedResponse(httpResponse, plainTextData));
+                } else {
+                    return Mono.just(httpResponse);
+                }
+            });
+        } else {
+            // Issue a get properties call to determine how much we will need to expand the range.
+            // Apply the request conditions from the download to the getProperties.
+            BlobRequestConditions rc = extractRequestConditionsFromRequest(requestHeaders);
+            return this.blobClient.getPropertiesWithResponse(rc).flatMap(response -> {
+                EncryptionData data = EncryptionData.getAndValidateEncryptionData(
+                    response.getValue().getMetadata().get(CryptographyConstants.ENCRYPTION_DATA_KEY),
+                    requiresEncryption);
+
+                // If there is no encryption data, then the blob is not encrypted. Pass the request through.
+                return data == null ? next.process() : Mono.just(data)
```

Comment:
Does this end up skipping the rest of the `HttpPipeline` when data isn't null?

Reply:
minor nit: may I suggest making this into an if-return block for data == null

```java
if (data == null) {
  return next.process();
}

return //decryption logic
```

as the else in the ternary operator is many lines

Reply:
We still call next.process() in the else block, so the rest of the pipeline shouldn't get skipped. I can refactor it as suggested.